### ISSUE TITLE
OS X Sierra compiling problem

### ISSUE THREAD

**Issue by totolol123** [otland/forgottenserver#1972]
```shell
[  1%] Generating CXX prefix source cotire/tfs_CXX_prefix.cxx
[  2%] Generating CXX prefix header cotire/tfs_CXX_prefix.hxx
[  3%] Building CXX precompiled header cotire/tfs_CXX_prefix.hxx.pch
In file included from .../build/cotire/tfs_CXX_prefix.hxx:4:
In file included from .../build/cotire/tfs_CXX_prefix.cxx:4:
In file included from .../src/otpch.h:42:
In file included from /usr/local/include/boost/asio.hpp:21:
In file included from /usr/local/include/boost/asio/basic_datagram_socket.hpp:20:
In file included from /usr/local/include/boost/asio/basic_socket.hpp:20:
In file included from /usr/local/include/boost/asio/basic_io_object.hpp:19:
In file included from /usr/local/include/boost/asio/io_service.hpp:767:
In file included from /usr/local/include/boost/asio/impl/io_service.hpp:71:
In file included from /usr/local/include/boost/asio/detail/task_io_service.hpp:196:
In file included from /usr/local/include/boost/asio/detail/impl/task_io_service.hpp:19:
In file included from /usr/local/include/boost/asio/detail/completion_handler.hpp:20:
In file included from /usr/local/include/boost/asio/detail/fenced_block.hpp:24:
/usr/local/include/boost/asio/detail/macos_fenced_block.hpp:45:5: error: 'OSMemoryBarrier' is deprecated: first deprecated in macOS 10.12 - Use std::atomic_thread_fence() from <atomic> instead [-Werror,-Wdeprecated-declarations]
    OSMemoryBarrier();
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/libkern/OSAtomicDeprecated.h:749:9: note: 'OSMemoryBarrier' has been explicitly marked deprecated here
void    OSMemoryBarrier( void );
        ^
In file included from .../build/cotire/tfs_CXX_prefix.hxx:4:
In file included from .../build/cotire/tfs_CXX_prefix.cxx:4:
In file included from .../src/otpch.h:42:
In file included from /usr/local/include/boost/asio.hpp:21:
In file included from /usr/local/include/boost/asio/basic_datagram_socket.hpp:20:
In file included from /usr/local/include/boost/asio/basic_socket.hpp:20:
In file included from /usr/local/include/boost/asio/basic_io_object.hpp:19:
In file included from /usr/local/include/boost/asio/io_service.hpp:767:
In file included from /usr/local/include/boost/asio/impl/io_service.hpp:71:
In file included from /usr/local/include/boost/asio/detail/task_io_service.hpp:196:
In file included from /usr/local/include/boost/asio/detail/impl/task_io_service.hpp:19:
In file included from /usr/local/include/boost/asio/detail/completion_handler.hpp:20:
In file included from /usr/local/include/boost/asio/detail/fenced_block.hpp:24:
/usr/local/include/boost/asio/detail/macos_fenced_block.hpp:51:5: error: 'OSMemoryBarrier' is deprecated: first deprecated in macOS 10.12 - Use std::atomic_thread_fence() from <atomic> instead [-Werror,-Wdeprecated-declarations]
    OSMemoryBarrier();
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/libkern/OSAtomicDeprecated.h:749:9: note: 'OSMemoryBarrier' has been explicitly marked deprecated here
void    OSMemoryBarrier( void );
        ^
2 errors generated.
CMake Error at cmake/cotire.cmake:1783 (message):
  cotire: error 1 precompiling .../build/cotire/tfs_CXX_prefix.hxx.
Call Stack (most recent call first):
  cmake/cotire.cmake:3332 (cotire_precompile_prefix_header)


make[2]: *** [cotire/tfs_CXX_prefix.hxx.pch] Error 1
make[1]: *** [CMakeFiles/tfs.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by esfomeado** [comment]
https://github.com/chriskohlhoff/asio/issues/144

**Comment by mazen-mardini** [comment]
I can definitely confirm this.
`/usr/local/include/boost/asio/detail/macos_fenced_block.hpp:51:5: 'OSMemoryBarrier' is deprecated: first deprecated in macOS 10.12 - Use std::atomic_thread_fence() from <atomic> instead`

**OS X version:** 10.12.3
**Boost**
```boost: stable 1.63.0 (bottled), HEAD
Collection of portable C++ source libraries
https://www.boost.org/
/usr/local/Cellar/boost/1.63.0 (12,491 files, 397.1M) *
  Poured from bottle on 2017-02-04 at 16:11:47
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/boost.rb
==> Dependencies
Optional: icu4c ✔
==> Requirements
Optional: mpi ✘
==> Options
--c++11
	Build using C++11 mode
--with-icu4c
	Build regexp engine with icu support
--with-mpi
	Build with MPI support
--without-single
	Disable building single-threading variant
--without-static
	Disable building static library variant
--HEAD
	Install HEAD version
```

We need a temporary workaround until asio is actually fixed. I tried cmake flags `-Wno-error=deprecated -Wno-deprecated`, but those didn't help suppress the error.

Any other ideas?

**Comment by ranisalt** [comment]
@mazen-mardini the warning is called `deprecated-declarations`, not `deprecated`. Try `-Wno-error=deprecated-declarations` instead.

It won't be added to the repository as it's not TFS's problem and seems to be already being solved in the dependencies with error.

**Comment by mazen-mardini** [comment]
@ranisalt That did it! `-Wno-error=deprecated-declarations` it is then.

**Comment by mazen-mardini** [comment]
This problem seems to have been fixed, check out https://github.com/chriskohlhoff/asio/commit/ffc9dd443fc2ca1aece0bbda14f3c521ceeb7924. It just hasn't been released yet, so lets hope it will be in Boost 1.64.

**Comment by kornholi** [comment]
No longer an issue. Builds correctly on High Sierra with Boost 1.66.0.
